pep-0677.rst

```diff
@@ -635,24 +635,107 @@ callable types and ``=>`` for lambdas.
 Runtime Behavior
 ----------------
 
-Our tentative plan is that:
-
-- The ``__repr__`` will show an arrow syntax literal.
-- We will provide a new API where the runtime data structure can be
-  accessed in the same manner as the AST data structure.
-- We will ensure that we provide an API that is backward-compatible
-  with ``typing.Callable`` and ``typing.Concatenate``, specifically
-  the behavior of ``__args__`` and ``__parameters__``.
+The new AST nodes need to evaluate to runtime types, and we have two goals for the
+behavior of these runtime types:
+
+- They should expose a structured API that is descriptive and powerful
+  enough to be compatible with extending the type to include new features
+  like named and variadic arguments.
+- They should also expose an API that is fully backward-compatible with
+  ``typing.Callable``.
+
+Evaluation and Structured API
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We intend to create new builtin types to which the new AST nodes will
+evaluate, exposing them in the ``types`` module.
+
+Our plan is to expose a structured API as if they were defined as follows::
+
+    class CallableType:
+        is_async: bool
+        arguments: Ellipsis | tuple[CallableTypeArgument]
+        return_type:: Typing.type
+
+    class CallableTypeArgument:
+        kind: CallableTypeArgumentKind
+        annotation: typing.Type | typing.TypeVar | typing.ParamSpec
```

Comment:
Arguably we shouldn't bother constraining the type of this field. It should be treated the same as `X` in `list[X]`, which can actually be any object (only type checkers complain about `list[42]`).